Carbon.Redis/Extensions/SerializationExtensions.cs

```diff
@@ -1,13 +1,80 @@
 ﻿using System;
 using System.IO;
+using System.Runtime.InteropServices;
+using System.Runtime.Serialization;
 using System.Runtime.Serialization.Formatters.Binary;
+using System.Text;
+using System.Text.Json;
 
 namespace Carbon.Caching.Abstractions.Extensions
 {
     public static class SerializationExtensions
     {
-        [Obsolete("This conversion is deprecated as of dotnet 5 unless you demand especially in your csproj")]
-        public static byte[] ToByteArray(this object obj)
+        public static byte[] ToByteArray(this object obj, Serialization serializationType = Serialization.BinaryFormatter)
+        {
+            if (serializationType == Serialization.BinaryFormatter)
+                return BinaryFormatterSerializer(obj);
+            else if (serializationType == Serialization.Json)
+                return JsonBinarySerializer(obj);
+            else if (serializationType == Serialization.Protobuf)
+                return ProtobufSerializer(obj);
+            else
+                throw new NotImplementedException();
+        }
+        public static T FromByteArray<T>(this byte[] byteArray, Serialization serializationType = Serialization.BinaryFormatter) where T : class
+        {
+            if (serializationType == Serialization.BinaryFormatter)
+                return BinaryFormatterDeserializer<T>(byteArray);
+            else if (serializationType == Serialization.Json)
+                return JsonBinaryDeserializer<T>(byteArray);
+            else if (serializationType == Serialization.Protobuf)
+                return ProtobufDeserializer<T>(byteArray);
+            else
+                throw new NotImplementedException();
+        }
+
+        private static T ProtobufDeserializer<T>(byte[] byteArray)
+            where T : class
+        {
+            throw new NotImplementedException();
+        }
+
+        private static byte[] ProtobufSerializer(object obj)
+        {
+            throw new NotImplementedException();
+        }
+
+        private static T JsonBinaryDeserializer<T>(byte[] byteArray) 
+            where T : class
+        {
+            if (byteArray == null)
+            {
+                return default(T);
+            }
+            var objAsJsonString = System.Text.Encoding.UTF8.GetString(byteArray);
+
+            try
+            {
+                var objectReturn = JsonSerializer.Deserialize<T>(objAsJsonString);
+                return objectReturn;
+            }
+            catch(JsonException)
+            {
+                return default(T);
+            }
+        }
+
+        private static byte[] JsonBinarySerializer(object obj)
+        {
+            if (obj == null)
+            {
+                return null;
+            }
+            var serializedObj = JsonSerializer.Serialize(obj);
+            return Encoding.UTF8.GetBytes(serializedObj);
```

Comment:
won't it slow us down now that we are converting to binary in addition to serializing the object into a json string? have you tested its impact on performance?

Reply:
Redis hashes store binaries rather than strings. In order to replace this binary formatter with another approach, you should use either json/xml serialization following with a binary conversion, or other as-is binary approaches such as protobuf.

It doesn't affect the performance as much as binary formatter does, plus security issues.

Reply:
have we discarded the idea of redis string storage instead of hashes then?

Reply:
Not discarded, but just replacing the existing one which is inherited from IDistributedCache interface using the hashes.

You can also use StringGet or Set method out of stack exchange. But using hash benefits from many aspect.